Library/ViewModels/NoShippingPostCampaignCheckoutViewModel.swift

```diff
@@ -290,21 +290,14 @@ public class NoShippingPostCampaignCheckoutViewModel: NoShippingPostCampaignChec
       .takeWhen(self.goToLoginSignupSignal)
       .map { project, reward in (LoginIntent.backProject, project, reward) }
 
-    // MARK: - Validate Checkout Details On Submit
+    // MARK: Create Payment Intent
 
-    let selectedCard = self.creditCardSelectedProperty.signal.skipNil()
-
-    let processingViewIsHidden = MutableProperty<Bool>(true)
-
-    // MARK: Validate Existing Cards
-
-    let newPaymentIntentForExistingCards = Signal.combineLatest(
+    let paymentIntentEvent = Signal.combineLatest(
       project,
       pledgeTotal,
       checkoutId,
       backingId
     )
```

Comment:
Just getting my brain back into things - is one of these linked to a `viewDidLoad` event? It would be good to guarantee that this can still only be called once per page load, and none of these properties make that obvious.

Reply:
All 4 of these fields should only have one value; `checkoutId`/`backingId` are generated by createCheckout (which is called on load) and `pledgeTotal` is calculated on load as well. However, this class is a bit of a mess- `pledgeTotal` here is also parsed from the backing, if there is any (which there can't be as it's a late pledge...), so it's definitely not clear from the signals that (if) `pledgeTotal` is only set once. So this should happen exactly once, but I agree that it'd be worth making that more clear and cleaning this class up a bit.

At a minimum, we may want to add some `skipNil`s (if needed - I can't tell from a quick look at the code if these are nullable) and a `takeFirst`, either to this signal or to its inputs.

Reply:
I like the idea of adding a `takeFirst`, just to make it very explicit!

Reply:
None of these signals emit optional values so I don't think we need any `skipNil`s here. I'm not super familiar with how `.take(first:)` functions, but adding it directly to this signal blocks it from completing. 

Do either of you know more about why that is? `checkoutId/backingId` rely on the CreateCheckout API call and the others are parsed out from the `PledgeViewData` that this VM is configured with. I'm not seeing where this would be called multiple times.

Reply:
Looks like `.take(first:)` means
> Take up to `n` values from the signal and then complete.
so that should be doing what we want (unless the completion itself is an issue). What happens if you try taking the first 2 or 3 etc?

Reply:
That's weird. Does `take(first: 1)` not work but `take(first: 2)` does? That definitely sounds like something getting called twice that shouldn't be (though I agree with Amy that it's possible that completing the signal itself is causing issues somehow). Let me know if you want help debugging this! I definitely don't think we should leave a `2` here without a comment explaining why; that's more misleading than not including a `take(first:)` at all.

Reply:
Also, my [test](https://github.com/kickstarter/ios-oss/pull/2210/commits/8f1dcc3a073fbef25d453e72f30dfaeff5157f97) asserts that only one payment intent request is being emitted. I don't see why `take(first: 1)` isn't working, tbh. My breakpoints/debugger aren't telling me much, unfortunately.

Reply:
I logged the events to look at them and here's what I think happens: `take(first: 1)` does exactly what it says it will, but the signal is completed, terminated, and disposed after. So the client secret is fetched and does exist, but it doesn't stick around. So in the existing card flow, for example, we see
```
let paymentIntentClientSecretForExistingCards = paymentIntentClientSecret
      .takeWhen(selectedCard)
```
but the `paymentIntentClientSecret` signal doesn't exist anymore when that card is selected. (I looked this up and `take(first:}` literally works by sending the completion event early.

I'm not sure if there's something similar to `take(first: 1)` that we can use instead that doesn't complete, but I definitely didn't find anything, so I think this is fine without it.